src/metabase/task/send_pulses.clj

```diff
@@ -1,136 +1,195 @@
-   (log/info "Sending scheduled pulses...")
-   (let [pulse-id->channels (group-by :pulse_id (pulse-channel/retrieve-scheduled-channels hour weekday monthday monthweek))]
-     (doseq [[pulse-id channels] pulse-id->channels]
-       (try
-         (task-history/with-task-history {:task         "send-pulse"
-                                          :task_details {:pulse-id pulse-id}}
-           (log/debugf "Starting Pulse Execution: %d" pulse-id)
-           (when-let [pulse (pulse/retrieve-notification pulse-id :archived false)]
-             (metabase.pulse/send-pulse! pulse :channel-ids (map :id channels)))
-           (log/debugf "Finished Pulse Execution: %d" pulse-id))
-         (catch Throwable e
-           (on-error pulse-id e)))))))
-
-; Clearing pulse channels is not done synchronously in order to support undoing feature.
+;;; ------------------------------------------------ Job: SendPulse ----------------------------------------------------
+
+(def ^:private send-pulse-job-key (jobs/key "metabase.task.send-pulses.send-pulse.job"))
+
+(defn- send-pulse-trigger-key
+  [pulse-id schedule-map]
+  (triggers/key (format "metabase.task.send-pulse.trigger.%d.%s"
+                        pulse-id (-> schedule-map
+                                     u.cron/schedule-map->cron-string
+                                     (str/replace " " "_")))))
+
+(defn- send-pulse!
+  [pulse-id channel-ids]
+  (try
+    (task-history/with-task-history {:task         "send-pulse"
+                                     :task_details {:pulse-id pulse-id}}
+      (log/debugf "Starting Pulse Execution: %d" pulse-id)
+      (when-let [pulse (pulse/retrieve-notification pulse-id :archived false)]
+        (metabase.pulse/send-pulse! pulse :channel-ids channel-ids))
+      (log/debugf "Finished Pulse Execution: %d" pulse-id))
+    (catch Throwable e
+      (log/errorf e "Error sending Pulse %d to channel ids: %s" pulse-id (str/join ", " channel-ids)))))
+
+(mu/defn ^:private send-pulse-trigger
+  "Build a Quartz trigger to send a pulse to a list of channel-ids."
+  ^CronTrigger
+  [pulse-id     :- pos-int?
+   schedule-map :- map?
+   pc-ids       :- [:maybe [:set pos-int?]]]
+  (when (seq pc-ids)
+    (triggers/build
+     (triggers/with-identity (send-pulse-trigger-key pulse-id schedule-map))
+     (triggers/for-job send-pulse-job-key)
+     (triggers/using-job-data {"pulse-id"    pulse-id
+                               "channel-ids" (set pc-ids)})
+     (triggers/with-schedule
+       (cron/schedule
+        (cron/cron-schedule (u.cron/schedule-map->cron-string schedule-map))
+        ;; if we miss a sync for one reason or another (such as system being down) do not try to run the sync again.
+        ;; Just wait until the next sync cycle.
+        ;;
+        ;; See https://www.nurkiewicz.com/2012/04/quartz-scheduler-misfire-instructions.html for more info
+        (cron/with-misfire-handling-instruction-ignore-misfires))))))
```

Comment:
Based on your comment, I think you've misunderstood what this misfire instruction does because it contradicts with the information in the link. For reference the previous schedule was set here: https://github.com/metabase/metabase/blob/7f53ac9343bbf41941ffdbf8e23a9b8f574219ea/src/metabase/task/send_pulses.clj#L126-L135

According to [the docs](https://www.javadoc.io/static/org.quartz-scheduler/quartz/2.1.7/org/quartz/Trigger.html#MISFIRE_INSTRUCTION_IGNORE_MISFIRE_POLICY), the new instructions you've added here means the job will execute as many times as it misfires to catch up after a system failure. That's what the original instructions were meant to avoid. See [this PR](https://github.com/metabase/metabase/pull/9961) that introduced them. I think what we want to keep that behaviour.

Here's the relevant part of on the ignore misfires policy:

> if a trigger uses this instruction, and it has missed several of its scheduled firings, then several rapid firings may occur as the trigger attempt to catch back up to where it would have been. For example, a SimpleTrigger that fires every 15 seconds which has misfired for 5 minutes will fire 20 times once it gets the chance to fire.

Here's the relevant part of the [linked article](https://nurkiewicz.com/2012/04/quartz-scheduler-misfire-instructions.html#:~:text=Immediately%20executes%20first%20misfired%20execution%20and%20discards%20other%20(i.e.%20all%20misfired%20executions%20are%20merged%20together).%20Then%20back%20to%20schedule.%20No%20matter%20how%20many%20trigger%20executions%20were%20missed%2C%20only%20single%20immediate%20execution%20is%20performed.) on the policy we used before:

> Immediately executes first misfired execution and discards other (i.e. all misfired executions are merged together). Then back to schedule. No matter how many trigger executions were missed, only single immediate execution is performed.

So I think we want to keep the old misfire policy. If there's a misfire because the thread pool is full, the job will get executed, unless a misfire happens again, then that misfire is discarded.

Reply:
you're absolutely correct.  we should use `cron/with-misfire-handling-instruction-fire-and-proceed` like before.